broadcast/broadcast.go

```diff
@@ -84,23 +114,20 @@ func (s *Broadcaster[T]) serve(ctx context.Context) {
 }
 
 func (s *Broadcaster[T]) removeSubscriber(sub <-chan T) {
-	// Cannot close using a read-only channel or remove the channel from a map.
-	for listener := range s.listeners {
-		if listener == sub {
-			delete(s.listeners, listener)
-			close(listener)
-		}
+	if listener, ok := s.listeners[sub]; ok {
+		delete(s.listeners, sub)
+		close(listener)
 	}
 }
 
 func (s *Broadcaster[T]) closeAll() {
-	for listener := range s.listeners {
+	for _, listener := range s.listeners {
 		close(listener)
 	}
 }
 
-func (s *Broadcaster[T]) ensureBroadcast(ctx context.Context, val T) {
-	for listener := range s.listeners {
+func (s *Broadcaster[T]) broadcastOrWait(ctx context.Context, val T) {
+	for _, listener := range s.listeners {
```

Comment:
I think we have a potential priority inversion here. A slow listener may make all the others wait indefinitely for the "guaranteed" delivery of the message. Maybe we should spawn each send in its own goroutine?

Reply:
I think you have to trust your consumers when using Wait strategy. Even when using goroutines we would still have to wait for all of them to finish using a WaitGroup.
I tested the following and it had a significant performance loss for fast consumers:
```go
func (s *Broadcaster[T]) broadcastOrWait(ctx context.Context, val T) {
	var wg sync.WaitGroup
	wg.Add(len(s.listeners))
	for _, listener := range s.listeners {
		listener := listener
		go func() {
			select {
			case listener <- val:
			case <-ctx.Done():
			}
			wg.Done()
		}()
	}
	wg.Wait()
}
```

Throughput was cut to 1/3 for a heavily buffered single consumer broadcast (compare to the other graph in the comment):

![Broadcast throughput for different buffer sizes _Throughput = #broadcasts _ (#subscribers+1) (1)](https://user-images.githubusercontent.com/1653035/194284770-3115bdf6-ee7c-4b66-8d1d-fd631cbd2045.png)